### PR TITLE
(CDAP-876) Upgrade Stream file to v2 format

### PR DESCRIPTION
**Pull request by chtyim**
- Added support for uni timestamp that uses file close time as event timestamp
- Added support for default headers

### CODE REVIEW

File: cdap-common/src/main/java/co/cask/cdap/common/stream/StreamEventDataCodec.java [hunk 1]
```diff
@@ -88,28 +89,49 @@ public static void encode(StreamEventData data, Encoder encoder) throws IOExcept
...
    * @return A new instance of {@link co.cask.cdap.api.stream.StreamEventData}.
    * @throws IOException If there is any IO error during decoding.
    */
   public static StreamEventData decode(Decoder decoder) throws IOException {
+    return decode(decoder, ImmutableMap.<String, String>of());
+  }
+
+  /**
+   * Decodes from the given {@link Decoder} to reconstruct a {@link StreamEventData}. A set of headers
+   * is used as the default set of headers.
```
Comment (awholegunch) [inline cdap-common/src/main/java/co/cask/cdap/common/stream/StreamEventDataCodec.java]: The set of headers provided is used as the default set of headers.

File: cdap-common/src/main/java/co/cask/cdap/common/stream/StreamEventDataCodec.java [hunk 2]
```diff
@@ -88,28 +89,49 @@ public static void encode(StreamEventData data, Encoder encoder) throws IOExcept
...
+   *
+   * @param decoder Decoder to read data from.
```
Comment (awholegunch) [inline cdap-common/src/main/java/co/cask/cdap/common/stream/StreamEventDataCodec.java]:
````
@param decoder the decoder to read data from
````
\[common formatting issues]

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamDataFileWriter.java [hunk 3]
```diff
@@ -110,6 +103,15 @@ public StreamDataFileWriter(OutputSupplier<? extends OutputStream> eventOutputSu
...
   }
 
+  /**
+   * Constructs a new instance that writes to given outputs.
+   *
+   * @param eventOutputSupplier Provides {@link OutputStream} for writing events.
+   * @param indexOutputSupplier Provides {@link OutputStream} for writing index.
+   * @param indexInterval Time interval in milliseconds for emitting new index entry.
+   * @param properties Set of properties that will be stored as file properties.
+   * @throws IOException If there is error in preparing the output streams.
```
Comment (awholegunch) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamDataFileWriter.java]:
````
@param eventOutputSupplier the provider of the {@link OutputStream} for writing events
@param indexOutputSupplier the provider of the {@link OutputStream} for writing the index
@param indexInterval the time interval in milliseconds for emitting a new index entry
@param properties the property set that will be stored as file properties
@throws IOException if there is an error in preparing the output streams
````

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamEventBuffer.java [hunk 4]
```diff
@@ -108,18 +111,20 @@ long getEndPosition() {
    * Decodes a stream event from the buffer.
    *
    * @param timestamp timestamp of the {@link PositionStreamEvent} created
+   * @param defaultHeaders set of headers default to the stream event
```
Comment (awholegunch) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamEventBuffer.java]: the set of headers that will used as the default for the stream event

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamEventBuffer.java]: changed

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamDataFileWriter.java (hunk 3 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamDataFileWriter.java]: changed

File: cdap-common/src/main/java/co/cask/cdap/common/stream/StreamEventDataCodec.java (hunk 2 above)
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/stream/StreamEventDataCodec.java]: changed

File: cdap-common/src/main/java/co/cask/cdap/common/stream/StreamEventDataCodec.java (hunk 1 above)
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/stream/StreamEventDataCodec.java]: changed

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamDataFileReader.java [hunk 5]
```diff
@@ -427,12 +482,21 @@ private void readDataBlock(ReadFilter filter) throws IOException {
...
+
+    // Use the template timestamp if available
+    timestamp = eventTemplate.getTimestamp() >= 0 ? eventTemplate.getTimestamp() : timestamp;
     if (filter.acceptTimestamp(timestamp)) {
       streamEventBuffer.fillBuffer(eventInput, readLength());
       this.timestamp = timestamp;
       return;
     }
 
+    // If timestamp is not accepted and the timestamp comes from event template, then the whole file can be ipipped
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamDataFileReader.java]: what does ipipped mean? is that a typo?

**Comment by albertshau** [comment]
:+1:

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamDataFileReader.java (hunk 5 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamDataFileReader.java]: I think it is "skipped".

**Comment by chtyim** [comment]
Build passed already: https://builds.cask.co/browse/CDAP-DUT497-1

**Pull request merged**